### PR TITLE
Auth routes || Signup & Login MVP

### PR DESCRIPTION
**Pull request by chrisrw**
Signup and Login page mvp completed (including basic email/password validation, and error handling.) Signup reroutes to login on success and login reroutes to the home screen on success. 
Screenshot 2023-06-01 at 1 07 17 PM

### CODE REVIEW

File: assets/bg-image.png
Comment (eklemen) [inline assets/bg-image.png:1]: was this file not present already?

**Comment by chrisrw** [comment]
No. I added that using the splash page image that was already present.

File: codegen.yml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 overwrite: true
-schema: "http://localhost:3000/graphql"
+schema: "env"
```
Comment (eklemen) [inline codegen.yml]: as is, this would break the codegen as `"env"` doesnt point to an actual schema set. I think theres some options on how we can pull this in dynamically or programmatically. For the time being, can you revert this change and just leave the localhost url in there?

**Comment by chrisrw** [comment]
For reverting this one change, is it better to make the change and then
commit and push it to the repo, or is there a specific way to make a
specific revision without the extra commit?

File: src/navStacks/HomeStack/index.tsx [hunk 2]
```diff
@@ -61,111 +64,115 @@ const AuthStack = createNativeStackNavigator<any>();
...
   return (
+
     <AuthStack.Navigator
       screenOptions={{ headerShown: false }}
       initialRouteName="LoginScreen"
     >
       <AuthStack.Screen name="LoginScreen" component={Login} />
+      <AuthStack.Screen name="Register" component={Register} />
+      <AuthStack.Screen name="VerificationCode" component={VerificationCode} />
+      <AuthStack.Screen name="Home" component={Home} />
```
Comment (eklemen) [inline src/navStacks/HomeStack/index.tsx]: We don't need to add Home to the AuthStack, it should already exist in the `MainBottomNavScreens`. We can just route the user to MainBottomNavScreens via `navigator.navigate('MainStack')` The MainStack has a prop to determine the default route, and we can add other logic in there too as needed.


Here is a crude representation of how nav stacks and tab navs are combined. A nav stack is a linear set of routes. A tab navigator adds a special bottom bar with icons and whatnot to route you around the app. Each of those tab navigators though can either be a single page or their own set of sub pages (a stack nav)
image

File: src/screens/Login/index.tsx [hunk 3]
```diff
@@ -1,15 +1,123 @@
...
+    try {
+      const res = await login({
+        variables: {
+          input: {
+            email: data.email,
+            password: data.password,
+          },
+        },
+      });
+      console.log(res.data?.login.data, 'data Response') // Data response object
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: i love that you commented what each step is doing, but it really isnt necessary here :)

My rule of thumb is to comment if there's a complex bit of logic or something that is hard to follow in order to explain what/why.

Also, remove the `console.log` :trollface:

File: src/screens/Login/index.tsx [hunk 4]
```diff
@@ -1,15 +1,123 @@
...
+      
+      const { refreshToken, idToken } = res.data?.login.data // Get tokens from Response
+      
+      if (idToken && refreshToken) {
+        await SecureStore.setItemAsync('idToken', idToken); // Store the tokens in SecureStore       
+        await SecureStore.setItemAsync('refreshToken', refreshToken);
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: we can wrap these 2 calls in a Promise.all so they run in parallel, like this:
```js
await Promise.all([
  SecureStore.setItemAsync('idToken', idToken), 
  SecureStore.setItemAsync('refreshToken', refreshToken)
]
```

File: src/screens/Login/index.tsx [hunk 5]
```diff
@@ -1,15 +1,123 @@
...
+        setErrorMessage('') // Clear error message
+        navigation.navigate('Home'); // Navigate to the home screen on success
+      }
+    } catch (error) {
+      console.log(error,'rrorrrr');
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: no `console.log` :trollface:

File: src/screens/Login/index.tsx [hunk 6]
```diff
@@ -1,15 +1,123 @@
...
+                rules={{
+                  required: { value: true, message: 'Password is required' },
+                  pattern: {
+                    value: passwordPattern,
+                    message: 'Password must contain at least 8 characters, including at least one uppercase letter, one lowercase letter, and one number',
+                  },
+                }}
+              />
+              <Link _text={{ fontSize: 'xs', fontWeight: '500', color: 'indigo.500' }} alignSelf="flex-end" mt="1">
+                Forget Password?
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: Change to `"Forgot Password"`

File: src/screens/Login/index.tsx [hunk 7]
```diff
@@ -1,15 +1,123 @@
...
+              </Link>
+              {errors.password && <FormControl.ErrorMessage>{errors.password.message}</FormControl.ErrorMessage>}
+            </FormControl>
+            {errorMessage ? <Text color="red.500">{errorMessage}</Text> : null}
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: Native base provides an error text component. Will that work instead of the `Text color red 500`?
```js
<FormControl.ErrorMessage _text={{
        fontSize: 'xs'
      }}>
```

File: src/screens/Login/index.tsx [hunk 8]
```diff
@@ -1,15 +1,123 @@
-import { Box, Column, Heading, Center } from 'native-base';
+import { Box, Heading, Center, FormControl, Input, Button, VStack, HStack, Link, Text } from 'native-base';
+import { useForm, Controller } from 'react-hook-form';
+import { useState } from 'react';
+import { useNavigation } from '@react-navigation/native';
+import { useLoginMutation, AuthUserInput } from '../../generated/graphql';
+import { TouchableOpacity } from 'react-native';
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: Lets use the Button component from native-base instead of TouchableOpacity from react-native

File: src/screens/Login/index.tsx [hunk 9]
```diff
@@ -1,15 +1,123 @@
...
+      <Center flex={1} py="10">
+        <Box w="90%">
+          <Heading size="lg" fontWeight="600" color="coolGray.800" _dark={{ color: 'warmGray.50' }}>
+            Welcome to CraftPile!
+          </Heading>
+          <Heading mt="1" _dark={{ color: 'warmGray.200' }} color="coolGray.600" fontWeight="medium" size="xs">
+            Sign in to continue!
+          </Heading>
+          <VStack space={3} mt="5">
+            <FormControl isInvalid={!!errors.email?.message}>
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: nit (nit pick) - a double bang `!!` coerces a value into a boolean. The safety operator `?.` will return either a value or `undefined` if the chain breaks. So these are a bit redundant together. The following should be sufficient.
```js
<FormControl isInvalid={errors?.email?.message}>
```

File: src/screens/Login/indexSave.tsx [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+        <FormControl.Label>Password</FormControl.Label>
+        <Controller
+          control={control}
+          name="password"
+          render={({ field: { onChange, onBlur, value } }) => (
+            <Input onChangeText={onChange} onBlur={onBlur} value={value} secureTextEntry />
+          )}
+          rules={{
+            required: { value: true, message: "Password is required" },
+            pattern: { value: passwordPattern, message: "Password must contain at least 8 characters, including at least one uppercase letter, one lowercase letter, and one number" }
```
Comment (eklemen) [inline src/screens/Login/indexSave.tsx:82]: for login, we dont really need to apply validation to their password. We already enforced that in registration, all we really care about is that the field is not blank, and of course that the email/pass is a valid login.

File: src/screens/Login/indexSave.tsx [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+export default Login;
+
+
+
+
+// import React from 'react';
+// import { Box, Heading, Center, FormControl, Input, Button, VStack, HStack, Link } from 'native-base';
+// import { useForm, Controller } from 'react-hook-form';
+// import { useNavigation } from '@react-navigation/native';
+// import { useLoginMutation, AuthUserInput } from '../../generated/graphql';
```
Comment (eklemen) [inline src/screens/Login/indexSave.tsx:117]: can remove dead/commented out code

File: src/screens/Login/indexSave.tsx [hunk 12]
```diff
@@ -0,0 +1,217 @@
+// This file is to save plain, known to work versions of the login component
```
Comment (eklemen) [inline src/screens/Login/indexSave.tsx]: Can delete this file as it is not doing anything.

File: src/screens/Login/index.tsx [hunk 13]
```diff
@@ -1,15 +1,123 @@
...
+                control={control}
+                name="password"
+                render={({ field: { onChange, onBlur, value } }) => (
+                  <Input onChangeText={onChange} onBlur={onBlur} value={value} secureTextEntry />
+                )}
+                rules={{
+                  required: { value: true, message: 'Password is required' },
+                  pattern: {
+                    value: passwordPattern,
+                    message: 'Password must contain at least 8 characters, including at least one uppercase letter, one lowercase letter, and one number',
```
Comment (eklemen) [inline src/screens/Login/index.tsx]: We don't need to validate the password input on Login since we already did in Registration. All we really care about here is that the field is not blank.

File: src/screens/Register/index.tsx [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+  const emailPattern = /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i;
+  const passwordPattern = /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*])[a-zA-Z0-9!@#$%^&*]{8,}$/;
+
+  const onSubmit = async (data: AuthUserInput) => {
+    const { email, password, passwordConfirmation } = data;
+
+    // Check if password fields match
+    if (password !== passwordConfirmation) {
+      Alert.alert("Passwords do not match");
```
Comment (eklemen) [inline src/screens/Register/index.tsx]: Lets not use Alert, can we just set validation error for the confirmation field?

File: src/screens/Register/index.tsx [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+    try {
+      const res = await register({
+        variables: {
+          input: {
+            email: email,
+            password: password,
+          },
+        },
+      });
```
Comment (eklemen) [inline src/screens/Register/index.tsx]: The way the the schemas are set up (and graphql generally goes) is that it will return a 200 if there are errors. A 200 signifies the server itself was able to process the request. In the backend im returning an object of `{ data, error }` or `{ success, error }` depending on the endpoint.
So the `catch` block is good to have, but we can also just use the `error` from `const [register, { data, loading, error }] = useRegistrationMutation();` and don't really need a setErrorMessage state hook.

In the case here, the try block will still execute if we get an error.
So instead we can base our logic off the following:
```js
error || data.error ? <error component> : null
```

**Review by eklemen — APPROVED** [review]
